### PR TITLE
Fix/actionmenu css prop type error

### LINKED ISSUE #1789: actionmenu: undesired css prop
Related to https://github.com/pluralsight/design-system/issues/1787

Need an approach that is comprehensive and feels like it resolves the root cause.

Need to understand how this emotion reference is making it into our code.

### PR DESCRIPTION
**Pull request by EdwardIrby**
resolves #1789 

Notes: https://docs.google.com/document/d/19FQJzP9QaHDRXYbjkhL0lR60sK5TGPAc4_f7DcEpmUo

Last two WIP commits are for components that use old Polymorphism approach and VerticalTabs and ActionMenu and If you look at the ActionMenu stories you'll see new polymorphic type error.

### CODE REVIEW

**Review by jaketrent — COMMENTED** [review]
Reviewed yesterday in person. Leaving note on decisions here:

1. Refactor HTMLPropsFor to use HTMLElement as generic type.
2. Refactor the couple of components that use an old polymorphic pattern to use the new polymorphic pattern which should be compatible with number 1.
3. Minimize work to ActionMenu due to imminent deprecation.

**Comment by EdwardIrby** [comment]
@jaketrent please re-review this and the linked doc that has more notes.

File: packages/util/src/types/index.ts
Comment (jaketrent) [inline packages/util/src/types/index.ts:3]: This abstraction has become an alias. Doesn't have to and maybe shouldn't happen this round, but I would like to remove this low-value alias and use the React type directly.

File: packages/table/src/react/__stories__/react-table.story.tsx
Comment (jaketrent) [reply packages/table/src/react/__stories__/react-table.story.tsx:762]: Confirmed, maybe? This story doesn't seem clickable. http://localhost:6006/?path=/story/components-table--row-selection But didn't cross-check with master.

**Review by jaketrent — COMMENTED** [review]
Looks great. Way to go. I have several comments that I'd like to have you reply to, please.

Builds, type checks, test. Spot-checked some stories.

File: packages/breadcrumb/tsconfig.json
Comment (EdwardIrby) [reply packages/breadcrumb/tsconfig.json:7]: Lack of strict on all the packages leads to typing errors. You shouldn't mix in match in this project.

File: packages/icon/src/react/__stories__/index.story.tsx
Comment (EdwardIrby) [reply packages/icon/src/react/__stories__/index.story.tsx:31]: It throws an error otherwise. If you pull down the branch and look at it you'll see and I honestly gave up trying to fix it after 15 minutes.

File: packages/table/src/react/__stories__/react-table.story.tsx
Comment (EdwardIrby) [reply packages/table/src/react/__stories__/react-table.story.tsx:762]: Can put it back. It's not part of the Table component. It's more something that should be in docs. Also was throwing type errors because of that and it wasn't typed properly to begin with.  Strict was not set to true the HTMLPropsFor refactor exposed underlying type issues as a result which is why I set it too true in all packages.

File: packages/util/src/types/index.ts
Comment (EdwardIrby) [reply packages/util/src/types/index.ts:3]: Easy to find and replace in this PR. You want to do it?

File: packages/banner/src/react/index.tsx
Comment (EdwardIrby) [reply packages/banner/src/react/index.tsx:70]: Yeah I recommend against it but this was one of those exceptions because it's wrapping a forwardRef Component the PSDS Button. Wouldn't of been so bad if you could import button without the forwardRef wrapping it.

File: packages/banner/src/react/index.tsx
Comment (EdwardIrby) [reply packages/banner/src/react/index.tsx:78]: Did you pull this down? I'm pretty sure that create a TS error but I could be wrong. I worked on this way too long resolving typescript issues. 

Honestly at the end I realized I personally rather use any for ref with React. Your call though if it does not throw an error. Just remember this is wrapping a component that if and Exotic one because it's wrapped in forwardRef already.

File: packages/button/src/react/__stories__/index.story.tsx
Comment (EdwardIrby) [reply packages/button/src/react/__stories__/index.story.tsx:218]: I felt the same but yea it's just more stuff. I also think we need to go and refactor primitives.ts and types.ts in utils because:
1. the arg pattern is counter to react in some of the Primitives with regards to forwardRef a
2. we need to settle on patterns for uses case not just use types.ts types that obfuscate what we want to to do and we forget how they do it later. We overload so much I found myself running into TS errors about overly complex types throughout this type fix refactoring task.

File: packages/button/src/react/index.tsx
Comment (EdwardIrby) [reply packages/button/src/react/index.tsx:179]: Because we often wrap Button in another component that is just passing some default props like in banner or error and then also forwardRef that component. In those use cases you have to use this more complex type. I would have done it for all components but I found it led to the original prop error inconsistently.  

Seriously I still don't see the pattern. I just know what prevents it most of the time, the new HTMLPropsFor alias. If you like try find and replacing in a few uses cases running builds and until the entire repo is converted. That's basically what I did for HTMLPropsFor when I wasn't fixing the type issues resulting from the refactor or lack of strict typing.

File: packages/button/src/react/index.tsx
Comment (EdwardIrby) [reply packages/button/src/react/index.tsx:220]: If it can be removed without throwing an error I'm for it.

File: packages/card/src/react/__specs__/index.spec.tsx
Comment (EdwardIrby) [reply packages/card/src/react/__specs__/index.spec.tsx:7]: why?

File: packages/card/src/react/__stories__/index.story.tsx
Comment (EdwardIrby) [reply packages/card/src/react/__stories__/index.story.tsx:42]: Naw I need to remove this. I created it before I pulled in a mod you made to the Card test.

File: packages/field/src/react/index.ts
Comment (EdwardIrby) [reply packages/field/src/react/index.ts:3]: I actually think we need to export this for all **Field** subcomponents the **typeof pattern** get's weird when it's working off a component that's wrapped in ForwardRef and we will need this interface building form inputs  off field.

File: packages/field/src/react/label.tsx
Comment (EdwardIrby) [reply packages/field/src/react/label.tsx:22]: I throws an error that's why. I don't know why since it actually exist on the HTMLLabelElement but it could have been vscode throwing a false flag. Try removing it in your env and running a build for field or I will tomorrow.

File: packages/banner/src/react/index.tsx
Comment (jaketrent) [reply packages/banner/src/react/index.tsx:78]: Yes, I pulled this down. I did not attempt to modify the code in this manner. 

I was drawn to this line because of the cast to `any`, because it signifies to me that it's a place where TS isn't working for us and we work around it. For that reason, I am currently not in favor of using an `any` cast as the default way of working with refs. But I'm open to hearing more about the reasoning that led you to that conclusion.

File: packages/button/src/react/__stories__/index.story.tsx
Comment (jaketrent) [reply packages/button/src/react/__stories__/index.story.tsx:218]: I agree with the "it's just more stuff" aspect of this potential refactor. I'd be grateful if you would form your suggestions for these refactors you mention into new Github issues so that they can be addressed later. Enough detail so they're dev-ready would be great. But if there's a conversation that needs to happen before a decision on approach, staging that conversation would be sufficient. Would you mind doing that, please?

File: packages/button/src/react/index.tsx
Comment (jaketrent) [reply packages/button/src/react/index.tsx:179]: This is unfortunate news. I think what it means is that we still don't understand the source of the problem or have full confidence in the solution. Things are likely much better after this PR, and we should go forward, but the issue still could come back. Or, also possible, the specific "this component is missing a required `css` prop" might still exist. Or, on that last point, did you actually test every component to ensure that didn't happen now? I did not.

File: packages/button/src/react/index.tsx
Comment (jaketrent) [reply packages/button/src/react/index.tsx:220]: Great, we are agreed. Do you plan to try removing it?

File: packages/card/src/react/__specs__/index.spec.tsx
Comment (jaketrent) [reply packages/card/src/react/__specs__/index.spec.tsx:7]: At this point the main reason is for codebase consistency. Currently, every other relative index import uses this syntax. 

Previously, as a part of the esm refactor, when we were going for true ES modules, the adjustment was made for the mandatory file extension https://nodejs.org/api/esm.html#esm_mandatory_file_extensions . I know we pulled a little back from that level of fidelity to the standard since then, and that's fine, but we did not remove this pattern.

File: packages/field/src/react/index.ts
Comment (jaketrent) [reply packages/field/src/react/index.ts:3]: What I think you're saying is it is not used today but that we're going to use it later as input components are refactored to use Field. Is that the case? 

If so, I think that's fine, but it seems like it doesn't belong in this PR because it's not cohesive with the purpose of the PR. Do you agree?

File: packages/util/src/types/index.ts
Comment (jaketrent) [reply packages/util/src/types/index.ts:3]: Um, sure, I will do it.

File: packages/banner/src/react/index.tsx
Comment (EdwardIrby) [reply packages/banner/src/react/index.tsx:78]: It does throw a type error. I can try and refactor button to use the polymorphic util `forwardRefWithAsAndStatics` and see if we can then use this typing. 

Type 'Ref<HTMLButtonElement | HTMLAnchorElement>' is not assignable to type 'LegacyRef<HTMLButtonElement> | undefined'.
  Type 'RefObject<HTMLButtonElement | HTMLAnchorElement>' is not assignable to type 'LegacyRef<HTMLButtonElement> | undefined'.
    Type 'RefObject<HTMLButtonElement | HTMLAnchorElement>' is not assignable to type 'RefObject<HTMLButtonElement>'.
      Type 'HTMLButtonElement | HTMLAnchorElement' is not assignable to type 'HTMLButtonElement'.
        Type 'HTMLAnchorElement' is missing the following properties from type 'HTMLButtonElement': disabled, form, formAction, formEnctype, and 11 more.ts(2322)
index.d.ts(147, 9): The expected type comes from property 'ref' which is declared here on type 'IntrinsicAttributes & ButtonProps'
(JSX attribute) ref?: React.LegacyRef<HTMLButtonElement> | undefined

File: packages/button/src/react/__stories__/index.story.tsx
Comment (EdwardIrby) [reply packages/button/src/react/__stories__/index.story.tsx:218]: On it!

File: packages/button/src/react/index.tsx
Comment (EdwardIrby) [reply packages/button/src/react/index.tsx:179]: Yep lots and lots and lots and lots and lots of builds for each component. Thus my mood lately

File: packages/field/src/react/index.ts
Comment (EdwardIrby) [reply packages/field/src/react/index.ts:3]: Sure I'll take it out and create a card linked to this ticket. I think I was using it but realized I could type better and left it there.

File: packages/util/src/types/index.ts
Comment (EdwardIrby) [reply packages/util/src/types/index.ts:3]: Maybe don't I'll take take care of it. It's actually a little more difficult because we use that alias in things other than interfaces. We also use it in type casting a numerous places. I'll do it if you haven't started.

Comment (jaketrent) [reply packages/util/src/types/index.ts:3]: Ok, I've stopped.

File: packages/button/src/react/index.tsx
Comment (EdwardIrby) [reply packages/button/src/react/index.tsx:220]: Removing it throws the following 

Type 'HTMLButtonElement | undefined' is not assignable to type 'HTMLButtonElement'.
  Type 'undefined' is not assignable to type 'HTMLButtonElement'.ts(2322)
index.d.ts(1095, 79): The expected type comes from the return type of this signature.
(property) MutableRefObject<HTMLButtonElement | undefined>.current: HTMLButtonElement | undefined

---
Currently we cast like so in the following
https://github.com/pluralsight/design-system/blob/15887e8fe9de05d2aef21cdf7240dc3097f50472/packages/dropdown/src/react/item.tsx#L35-L40
https://github.com/pluralsight/design-system/blob/15887e8fe9de05d2aef21cdf7240dc3097f50472/packages/emptystate/src/react/index.tsx#L88-L90
https://github.com/pluralsight/design-system/blob/15887e8fe9de05d2aef21cdf7240dc3097f50472/packages/errors/src/react/error-page.tsx#L93-L95
https://github.com/pluralsight/design-system/blob/15887e8fe9de05d2aef21cdf7240dc3097f50472/packages/focusmanager/src/react/index.tsx#L14-L16
https://github.com/pluralsight/design-system/blob/15887e8fe9de05d2aef21cdf7240dc3097f50472/packages/position/src/react/index.tsx#L103-L105
https://github.com/pluralsight/design-system/blob/15887e8fe9de05d2aef21cdf7240dc3097f50472/packages/radio/src/react/button.tsx#L63-L66
https://github.com/pluralsight/design-system/blob/15887e8fe9de05d2aef21cdf7240dc3097f50472/packages/scrollable/src/react/index.tsx#L55-L57
https://github.com/pluralsight/design-system/blob/15887e8fe9de05d2aef21cdf7240dc3097f50472/packages/switch/src/react/index.tsx#L132-L137

Do you want to create card to figure out a way to remove all these as a separate issue? This popped up in button now because of the `strict:true`.

File: packages/button/src/react/index.tsx
Comment (EdwardIrby) [reply packages/button/src/react/index.tsx:179]: On that note should we create a card to ideate how we might automate testing this so we don't have to manually do this or wait for an issue to be filed again? It's happened repeatedly over the last 3-4 months.

Comment (jaketrent) [reply packages/button/src/react/index.tsx:179]: ^ card for automated test ideas works for me.

**Review by jaketrent — APPROVED** [review]
I'm good to merge if you think it is.

File: packages/banner/src/react/index.tsx
Comment (EdwardIrby) [reply packages/banner/src/react/index.tsx:80]: Going to make this a separate issue

**Pull request merged**
